### PR TITLE
Segwit

### PR DESCRIPTION
**Pull request by xenog**
Support for segwit. Please review thoroughly. Comments and modifications strongly encouraged.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
We can probably skip the "segnet" and "segwit addresses" commits - since testnet3 now has segwit no need for segnet, and the BIP for segwit addresses was withdrawn pending future work in the space.

**Comment by TheBlueMatt** [comment]
@schildbach Please, please, do not encourage people to add bitcoinj-specific addresses for transaction types. There isnt any rush to have addresses if no one else will support them - skip it for now and wait a month or two until there are other proposals and the community can have a reasonable discussion.

**Comment by schildbach** [comment]
All I wanted to say is: I see no need to remove that implementation that has alredy been written. Don't worry we'll set no precedent here: I do not intend to merge unless it's clear segwit activates, so there is plenty of time for more proposals/implementations.

**Comment by TheBlueMatt** [comment]
Mmm, ok, I mean I would still remove the address format so that it is ready to merge if segwit were to activate (as it would still otherwise set precedent), but, I suppose its a moot point unless segwit gets much closer.

**Comment by rktoomey** [comment]
It would be helpful to have this available as soon as possible because right now bitcoinj is not working with testnet3 on bitcoind 0.13+ nodes - see `0000000000332ef392f66e80afa2bb1389c08f47050ea189ae432e7bfea8c36e` for an example of a block that fails to successfully deserialize.  .

**Comment by schildbach** [comment]
@rktoomey Segwit is a softfork, so if current bitcoinj fails with a segwit block, that's a serious bug. Can you open a separate issue and attach the logfile?

**Comment by schildbach** [comment]
@rktoomey Is your app using a bloom filter? I had a look at my Bitcoin Wallet logfile and it had no trouble synching past the block. But that's expected, as it hardly sees any transactions due to the filter.

**Comment by rktoomey** [comment]
@schildbach I'm filing an issue with exact details - there are multiple factors.  Users who are using pre 0.13 nodes should be unaffected.

**Comment by Polve** [comment]
It seems like in a way or another we are getting segwit: could this be merged?

**Comment by xenog** [comment]
I'm doing some cleanup at the moment, merging latest master in, removing Base58 SegWit addresses, looking at block validation and P2P support. I'll keep you posted.

**Comment by xenog** [comment]
I changed base of this commit to the `segwit` branch as recommended.

**Comment by xenog** [comment]
GitHub went bananas when I rebased.

**Comment by xenog** [comment]
I can't reopen this PR. I'll create a new one.

**Comment by luke-jr** [comment]
@xenog Are you working on versionbits support?

**Comment by xenog** [comment]
@luke-jr if you mean BIP-9 support, I haven't started yet. Want to help?

File: examples/src/main/java/org/bitcoinj/examples/SimpleTransaction.java [hunk 1]
```diff
@@ -1,50 +0,0 @@
-/*
```
Comment (schildbach) [inline examples/src/main/java/org/bitcoinj/examples/SimpleTransaction.java]: You removed the example again, was this intentional?

**Comment by schildbach** [comment]
Merged this to segwit. Code style looks very good, and thanks for all the tests! I didn't review all the logic yet, guess I have to learn a lot about segwit (-:

If you see on the segwit branch a refactoring that would benefit bitcoinj already now, and works towards merging the branch, don't hesitate to mention this because merging and testing "preparational" changes early helps kepping the impact of merging the segwit feature low.

**Comment by schildbach** [comment]
Yes, if anyone wants to tackle BIP9 (on a separate PR), that'll be great. It's a prerequisite for proper segwit support. Also, I think Bech32/segwit addresses hasn't been implemented yet.

**Comment by xenog** [comment]
No objections here. I am working on consensus code at the moment. On my roadmap for the next week or two:

- Counting segwit sigops
- Verifying witness structure in coinbase when segwit is active
- Peer-to-peer support for moving segwit data

**Comment by tiger1342** [comment]
Hello guy,

I am trying to use Segwit in BitcoinJ but didn't find anything about it in API. please let me know if it was ever committed into Master?

**Pull request merged**